File: lib/id3.js
// Минимальный ID3v2.3 writer для embed обложки + базовых тегов в MP3.
// Достаточно для того чтобы Winamp/foobar/AIMP/VLC/Apple Music/Windows Media
// показывали title/artist/album/cover.
//
// Структура ID3v2.3:
//   "ID3" + ver(0x03,0x00) + flags(1B) + size(4B syncsafe)
//   + frames: TIT2/TPE1/TALB/TDRC (text) и APIC (cover)
//
// Используется в background.js — Web Crypto API + TextEncoder доступны там.

(function () {
  function concat(...arrs) {
    let total = 0;
    for (const a of arrs) total += a.length;
    const out = new Uint8Array(total);
    let off = 0;
    for (const a of arrs) { out.set(a, off); off += a.length; }
    return out;
  }

  function be32(n) {
    return new Uint8Array([(n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff]);
  }

  function syncsafe32(n) {
    return new Uint8Array([(n >> 21) & 0x7f, (n >> 14) & 0x7f, (n >> 7) & 0x7f, n & 0x7f]);
  }

  function asciiBytes(str) {
    const out = new Uint8Array(str.length);
    for (let i = 0; i < str.length; i++) out[i] = str.charCodeAt(i) & 0x7f;
    return out;
  }

  // UTF-16LE с BOM + двойной нулевой терминатор. В ID3v2.3 это encoding 0x01.
  // ID3v2.3 НЕ поддерживает UTF-8 (0x03) — Windows Explorer такие теги
  // игнорирует. UTF-16 — единственный способ записать кириллицу в 2.3.
  function utf16leWithBom(str) {
    const s = String(str);
    const out = new Uint8Array(2 + s.length * 2 + 2);
    out[0] = 0xff; out[1] = 0xfe;  // BOM (little-endian)
    let p = 2;
    for (let i = 0; i < s.length; i++) {
      const c = s.charCodeAt(i);
      out[p++] = c & 0xff;
      out[p++] = (c >> 8) & 0xff;
    }
    out[p++] = 0x00; out[p++] = 0x00;  // terminator
    return out;
  }

  function textFrame(id, text) {
    if (!text) return new Uint8Array(0);
    // encoding 0x01 = UTF-16 with BOM (валидно в ID3v2.3)
    const body = concat(new Uint8Array([0x01]), utf16leWithBom(text));
    const head = concat(asciiBytes(id), be32(body.length), new Uint8Array([0, 0]));
    return concat(head, body);
  }

  function apicFrame(imageBytes, mime) {
    // APIC: encoding 0x00 (ISO-8859-1) — description пустой, mime это ASCII,
    // так максимально совместимо. Picture data от encoding не зависит.
    const body = concat(
      new Uint8Array([0x00]),                  // text encoding ISO-8859-1
      asciiBytes(mime || 'image/jpeg'),        // MIME (ASCII)
      new Uint8Array([0x00]),                  // MIME terminator
      new Uint8Array([0x03]),                  // picture type: 3 = cover (front)
      new Uint8Array([0x00]),                  // empty description terminator
      imageBytes,
    );
    const head = concat(asciiBytes('APIC'), be32(body.length), new Uint8Array([0, 0]));
    return concat(head, body);
  }

  // USLT — несинхронный текст песни. body: [enc][lang 3][descriptor+term][lyrics]
  function usltFrame(lyrics) {
    // encoding 0x01 (UTF-16) для кириллицы. descriptor пустой = BOM+терминатор.
    const body = concat(
      new Uint8Array([0x01]),                 // UTF-16 encoding
      asciiBytes('rus'),                       // language (ISO-639-2)
      new Uint8Array([0xff, 0xfe, 0x00, 0x00]),// пустой descriptor: BOM + UTF-16 terminator
      utf16leWithBom(lyrics),                  // сам текст
    );
    const head = concat(asciiBytes('USLT'), be32(body.length), new Uint8Array([0, 0]));
    return concat(head, body);
  }

  function buildID3v2(opts) {
    // opts: { title, artist, album, year, cover (Uint8Array), coverMime, lyrics }
    const frames = [];
    if (opts.title)  frames.push(textFrame('TIT2', opts.title));
    if (opts.artist) frames.push(textFrame('TPE1', opts.artist));
    if (opts.album)  frames.push(textFrame('TALB', opts.album));
    if (opts.year)   frames.push(textFrame('TDRC', String(opts.year)));
    if (opts.lyrics) frames.push(usltFrame(opts.lyrics));
    if (opts.cover)  frames.push(apicFrame(opts.cover, opts.coverMime || 'image/jpeg'));

    const body = concat(...frames);
    const enc = new TextEncoder();
    const header = concat(
      enc.encode('ID3'),
      new Uint8Array([0x03, 0x00]),  // version 2.3.0
      new Uint8Array([0x00]),         // flags
      syncsafe32(body.length),
    );
    return concat(header, body);
  }

  function stripExistingID3v2(mp3Bytes) {
    // Если в начале mp3 уже есть ID3v2 — вырезаем, чтобы не было дубликатов
    if (mp3Bytes.length < 10) return mp3Bytes;
    if (mp3Bytes[0] === 0x49 && mp3Bytes[1] === 0x44 && mp3Bytes[2] === 0x33) {
      const size = ((mp3Bytes[6] & 0x7f) << 21) |
                   ((mp3Bytes[7] & 0x7f) << 14) |
                   ((mp3Bytes[8] & 0x7f) << 7) |
                    (mp3Bytes[9] & 0x7f);
      return mp3Bytes.subarray(10 + size);
    }
    return mp3Bytes;
  }

  function prependID3v2ToMP3(mp3Bytes, tagBytes) {
    const clean = stripExistingID3v2(mp3Bytes);
    return concat(tagBytes, clean);
  }

  globalThis.YMD_ID3 = { buildID3v2, prependID3v2ToMP3 };
})();

File: android/app/src/main/java/ru/ymd/downloader/Id3.kt
package ru.ymd.downloader

import java.io.ByteArrayOutputStream

/**
 * Запись тегов ID3v2.3 в MP3: название, исполнитель, альбом, жанр, номер трека,
 * текст песни и обложка.
 *
 * Важная деталь, проверенная на практике: в ID3v2.3 кодировка UTF-8 недопустима.
 * Текстовые поля пишем в UTF-16 с BOM, иначе проводник Windows молча
 * игнорирует такие теги вместе с обложкой.
 *
 * Порт реализации из расширения (lib/id3.js).
 */
object Id3 {

    private fun ascii(s: String): ByteArray = s.toByteArray(Charsets.US_ASCII)

    private fun be32(n: Int): ByteArray = byteArrayOf(
        ((n ushr 24) and 0xff).toByte(), ((n ushr 16) and 0xff).toByte(),
        ((n ushr 8) and 0xff).toByte(), (n and 0xff).toByte()
    )

    /** Размер тега пишется «синхробезопасно»: по 7 бит в байте. */
    private fun syncsafe32(n: Int): ByteArray = byteArrayOf(
        ((n shr 21) and 0x7f).toByte(), ((n shr 14) and 0x7f).toByte(),
        ((n shr 7) and 0x7f).toByte(), (n and 0x7f).toByte()
    )

    /** UTF-16LE с BOM и двойным нулём в конце. */
    private fun utf16(s: String): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(0xff); out.write(0xfe)
        for (ch in s) {
            val c = ch.code
            out.write(c and 0xff)
            out.write((c shr 8) and 0xff)
        }
        out.write(0); out.write(0)
        return out.toByteArray()
    }

    private fun textFrame(id: String, text: String): ByteArray {
        if (text.isBlank()) return ByteArray(0)
        val body = ByteArrayOutputStream()
        body.write(0x01)              // 0x01 — UTF-16, допустимо в ID3v2.3
        body.write(utf16(text))
        val b = body.toByteArray()
        val out = ByteArrayOutputStream()
        out.write(ascii(id)); out.write(be32(b.size)); out.write(0); out.write(0)
        out.write(b)
        return out.toByteArray()
    }

    /** Текст песни: язык + пустое описание + сам текст. */
    private fun usltFrame(lyrics: String): ByteArray {
        if (lyrics.isBlank()) return ByteArray(0)
        val body = ByteArrayOutputStream()
        body.write(0x01)
        body.write(ascii("rus"))
        body.write(byteArrayOf(0xff.toByte(), 0xfe.toByte(), 0, 0))  // пустое описание
        body.write(utf16(lyrics))
        val b = body.toByteArray()
        val out = ByteArrayOutputStream()
        out.write(ascii("USLT")); out.write(be32(b.size)); out.write(0); out.write(0)
        out.write(b)
        return out.toByteArray()
    }

    /** Обложка. Кодировку берём ISO-8859-1: описание пустое, MIME — чистый ASCII. */
    private fun apicFrame(cover: ByteArray, mime: String): ByteArray {
        val body = ByteArrayOutputStream()
        body.write(0x00)
        body.write(ascii(mime)); body.write(0)
        body.write(0x03)              // 3 — обложка (передняя сторона)
        body.write(0)                 // описание пустое
        body.write(cover)
        val b = body.toByteArray()
        val out = ByteArrayOutputStream()
        out.write(ascii("APIC")); out.write(be32(b.size)); out.write(0); out.write(0)
        out.write(b)
        return out.toByteArray()
    }

    fun buildTag(track: Track, cover: ByteArray?): ByteArray {
        val frames = ByteArrayOutputStream()
        frames.write(textFrame("TIT2", track.title))
        frames.write(textFrame("TPE1", track.artist))
        frames.write(textFrame("TALB", track.album))
        frames.write(textFrame("TPE2", track.albumArtist))
        frames.write(textFrame("TDRC", track.year))
        frames.write(textFrame("TCON", track.genre))
        if (track.trackNum > 0) {
            val v = if (track.trackTotal > 0) "${track.trackNum}/${track.trackTotal}"
            else track.trackNum.toString()
            frames.write(textFrame("TRCK", v))
        }
        if (track.discNum > 0) frames.write(textFrame("TPOS", track.discNum.toString()))
        frames.write(usltFrame(track.lyrics))
        if (cover != null && cover.isNotEmpty()) frames.write(apicFrame(cover, "image/jpeg"))

        val body = frames.toByteArray()
        val out = ByteArrayOutputStream(body.size + 10)
        out.write(ascii("ID3"))
        out.write(0x03); out.write(0x00)   // версия 2.3.0
        out.write(0x00)                    // флаги
        out.write(syncsafe32(body.size))
        out.write(body)
        return out.toByteArray()
    }

    /** Убираем ранее приклеенный тег, чтобы не было дублей. */
    private fun stripExisting(mp3: ByteArray): ByteArray {
        if (mp3.size < 10) return mp3
        if (mp3[0] == 'I'.code.toByte() && mp3[1] == 'D'.code.toByte() && mp3[2] == '3'.code.toByte()) {
            val size = ((mp3[6].toInt() and 0x7f) shl 21) or
                    ((mp3[7].toInt() and 0x7f) shl 14) or
                    ((mp3[8].toInt() and 0x7f) shl 7) or
                    (mp3[9].toInt() and 0x7f)
            val start = 10 + size
            if (start in 0..mp3.size) return mp3.copyOfRange(start, mp3.size)
        }
        return mp3
    }

    fun attach(mp3: ByteArray, track: Track, cover: ByteArray?): ByteArray {
        val clean = stripExisting(mp3)
        val tag = buildTag(track, cover)
        val out = ByteArray(tag.size + clean.size)
        System.arraycopy(tag, 0, out, 0, tag.size)
        System.arraycopy(clean, 0, out, tag.size, clean.size)
        return out
    }
}
